Add Auth0 integration

bananaoomarang opened the pull request:
Adds support for securing webtasks with a JWT, authenticating webtasks with Lock. (https://github.com/auth0/wt-cli/issues/19).

woloski commented on bin/create.js:
```diff
@@ -263,6 +275,18 @@ function handleCreate (argv) {
+
+                if (argv.auth) {
+                    try {
+                        auth_token = jwt.sign({}, argv.auth);
```
why are you generating a token here?

bananaoomarang replied:
This is just to give the user something to call their webtask with, though it should give the generated secret too/optionally.

woloski commented:
I was expecting to see that when you specify 

```
wt create foo.js --auth --secret=h09fj2309d23h023je0
```

This would be like setting the secret in `ectx` under some convention (like JWT_SECRET) and then use that on the proxy to verify the token that the end user might send.

bananaoomarang commented:
I did start putting it under `ectx` first, but then I realised that if it goes over https to be saved in the proxy there shouldn't be any need for that.

I wasn't sure whether to expose setting your own secret, since my local version has support for specifying `--auth=<email/domain>` already. It might be a good idea to pull a secret (optionally) from `.env`, which would then in effect become a 'password'

woloski commented:
can you point me where you are storing the secret on the proxy?

bananaoomarang commented:
Currently it's embedded in the jtn token at `auth.secret`.

Or: if `auth.email` is set then we try to use the header token to grab `userinfo` from a shared Aith0 account (though there's some bugs I was working out with the latter, so that's not on this PR just yet)

ggoodman commented on bin/create.js:
```diff
@@ -139,15 +161,49 @@ module.exports = Cli.createCommand('create', 'Create webtasks', {
+                              .map(function (str) {
+                                  return str.trim();
+                              })
+                              .filter(function (str) {
```
Nice shortcut for this is `.filter(Boolean)`.

ggoodman commented on bin/create.js:
```diff
@@ -257,12 +314,16 @@ function handleCreate (argv) {
                                 + (argv.container || profile.container)
                                 + '/' + argv.name
-                                + (argv.prod ? '': '?webtask_no_cache=1');
+                                + (argv.prod ? '': '?webtask_no_cache=1')
```
Let's refactor this to something that isn't a bunch of nested ternaries.

ggoodman commented on index.js:
```diff
@@ -166,6 +165,12 @@ WebtaskProfile.prototype.createToken = Bluebird.method(function (options, cb) {
+
+    if (options.auth0)
+        params.auth = options.auth0;
+    else if (options.secure)
```
Shouldn't this be `options.share`? Otherwise where is `options.secure` coming from?

ggoodman commented on bin/create.js:
```diff
@@ -6,6 +6,7 @@ var Watcher = require('filewatcher');
 var Webtask = require('../');
 var _ = require('lodash');
 var crypto = require('crypto');
```
Can you caps this?

ggoodman commented on bin/create.js:
```diff
@@ -6,6 +6,7 @@ var Watcher = require('filewatcher');
 var Webtask = require('../');
 var _ = require('lodash');
 var crypto = require('crypto');
+var jwt = require('jsonwebtoken');
```
And this?

tjanczuk commented on index.js:
```diff
@@ -166,6 +165,9 @@ WebtaskProfile.prototype.createToken = Bluebird.method(function (options, cb) {
         params.jtn = options.name;
+
+    if (options.auth0)
+        params.auth = options.auth0;
```
This is going to result in a webtask token that contains the Auth0 client secret in plaintext. The Auth0 client secret must be stored as part of ectx.

bananaoomarang commented:
The only remaining issue at the moment is that you can't `curl mywebtask?key=abcd` directly yet, you have to do `curl mywebtask_auth0?key=abcd`. This is because the proxy sanitizes the requests of auth headers and query `key`s before proxying to the sandbox. Will look at solving that tomorrow morning.

bananaoomarang commented:
Just rebased to clean up the mess, but I have removed the shared account and in built email checking for simplicity, and added a new, simpler syntax.

```
$ wt create task --auth0 '<clientID> <client-secret> <auth0-domain>'
$ wt create task --auth0 (same as above, but pulls from env)
$ wt create task --share (as before, generates a 'shareable' URL)
```

tjanczuk commented on bin/create.js:
```diff
@@ -37,6 +50,10 @@ var tokenOptions = {
         type: 'string',
     },
+    auth0: {
+        description: 'Authenticate webtask with Auth0: --auth0 <clientID> <client-secret> <auth0-domain>',
```
This should show quotes around <clientID> <client-secret> <auth0-domain>. I wonder if a better pattern would be to use commas or separate options for each of these?

bananaoomarang replied:
They were previously separate options, but since you have to specify all three together anyway I thought they'd be better rolled into one. Will make them comma separated.

tjanczuk commented on bin/create.js:
```diff
@@ -57,6 +74,10 @@ var tokenOptions = {
         type: 'string',
     },
+    auth: {
+      description: 'set webtask permisions',
```
This is for authorization, right? We should be more descriptive with what people can say here, give an example or two, even if this is just an e-mail address or e-mail address domain.

tjanczuk replied:
Maybe we can call it `access` or `acls`. The `auth` is confusing vs `auth0` sitting next to it.

tjanczuk replied:
Is this the main switch that controls whether the webtask will require authentication, or something else? If so, the "set webtask permissions" does not express it.

bananaoomarang replied:
Think this used to be where you could specify emails and domains, you're right it's no longer needed.

tjanczuk commented on bin/create.js:
```diff
@@ -150,18 +167,36 @@ module.exports = Cli.createCommand('create', 'Create webtasks', {
+                        clientId: auth0Credentials[0]     || process.env.AUTH0_CLIENT_ID,
+                        clientSecret: auth0Credentials[1] || process.env.AUTH0_CLIENT_SECRET,
+                        domain: auth0Credentials[2]       || process.env.AUTH0_DOMAIN,
+                    };
```
So here is the problem with positional arguments: you can specify (clientId), (clientId, clientSecret), or (clientId, clientSecret, domain) only. Maybe having individual options for each would be beneficial?

bananaoomarang replied:
I'm not sure about this. Previously they were individual options, but since all three necessarily need to be specified together it adds verbosity in the user commands and the code. I think having `--auth0` is nice from a branding perspective as well, so I've left it for now.

tjanczuk commented on bin/create.js:
```diff
@@ -302,26 +337,76 @@ function handleCreate (argv) {
+                    var named_url = profile.url
+                        + '/api/run/'
+                        + (argv.container || profile.container)
+                        + '/' + argv.name;
```
This is just `unnamed_url + '/' + argv.name`, right?

tjanczuk commented on bin/create.js:
```diff
@@ -302,26 +337,76 @@ function handleCreate (argv) {
+                        + '/' + argv.name;
+
+                if(argv.share) {
+                    var aud = named_url.replace('https:\/\/', '');
```
If argv.share is present, is argv.name also ALWAYS present? If not, the named_url here will be undefined.

tjanczuk commented on bin/create.js:
```diff
@@ -302,26 +337,76 @@ function handleCreate (argv) {
+                        name: argv.name,
+                        code: Fs.readFileSync(__dirname + '/lib/lock_webtask.js', 'utf8'),
+                        param: {
+                            baseUrl: profile.url,
```
Isn't his overridable per `wt create`? it used to be that profile only provided defaults, but all properties could be specified explicitly on create. This was supposed to make scripting easy.

tjanczuk commented on bin/create.js:
```diff
@@ -364,6 +449,22 @@ function parseDate (argv, field) {
 
+function getSecret() {
+    try {
+        return (process.env.WEBTASK_JWT_SECRET || Crypto.randomBytes(128).toString('base64'));
```
Does generation of a random key only really make sense with `--share`? Are we doing this validation somewhere?

bananaoomarang replied:
I think it makes sense with `--share`, but you should probably be able to override it. It's validated in `coreos-tools`, this makes the given URL the only one that'll work unless you get the secret out of `wt-cli` and sign your own tokens.

tjanczuk commented on bin/create.js:
```diff
@@ -364,6 +449,22 @@ function parseDate (argv, field) {
+    try {
+        return (process.env.WEBTASK_JWT_SECRET || Crypto.randomBytes(128).toString('base64'));
+    } catch(e) {
+        console.log('Couldn\'t generate random secret: ' + e.message);
```
Should this just throw? In any way it probably shouldn't silently return undefined.

tjanczuk commented on package.json:
```diff
@@ -31,12 +31,15 @@
+    "dotenv": "^1.2.0",
     "filewatcher": "^3.0.0",
+    "jsonwebtoken": "^5.0.4",
     "jws": "^3.0.0",
```
Can we normalize on using one of jws or jsonwebtoken?

tjanczuk commented on bin/lib/lock_webtask.js:
```diff
@@ -0,0 +1,155 @@
+        <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=0" />
+        <link rel="shortcut icon" href="https://webtask.io/images/favicon/favicon.ico">
+
+        <title>Authenticate</title>
```
Custom title would be a nice touch.

tjanczuk commented on bin/lib/lock_webtask.js:
```diff
@@ -0,0 +1,155 @@
+
+    if(!ctx.data.container || !ctx.data.taskname) {
+       res.statusCode = 400;
+       res.end('Must provide container & taskname params');
```
This (and subsequent cases) should be `return res.end(...)`.

tjanczuk commented on bin/lib/lock_webtask.js:
```diff
@@ -0,0 +1,155 @@
+       res.end('Must provide container & taskname params');
+    }
+
+    if(!ctx.data.clientId && !ctx.data.domain) {
```
Should this be || instead of &&?

tjanczuk commented on bin/lib/lock_webtask.js:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    let query = qs.parse(
+        url.parse(req.url).query
```
I think you need to call `url.parse(req.url, true)` if you want the query parsed as well: https://nodejs.org/api/url.html#url_url_parse_urlstr_parsequerystring_slashesdenotehost

tjanczuk commented on bin/lib/lock_webtask.js:
```diff
@@ -0,0 +1,155 @@
+    'clientId',
+];
+
+app.get('*', (req, res) => {
```
Is there any value in pulling in express machinery if you have just one handler?

tjanczuk commented on bin/lib/lock_webtask.js:
```diff
@@ -0,0 +1,155 @@
+                REQ_METHOD: req.method,
+                CLIENT_ID: ctx.data.clientId,
+                AUTH0_DOMAIN: ctx.data.domain,
+                FUNC_TO_RUN,
```
Would it make sense to just embed func to run in the html? It does not depend on any client input.

bananaoomarang replied:
I think I prefer it outside of the HTML because it makes editing/maintaining easier. I gave it a better name, but if you feel strongly there's no reason it can't be embedded in the HTML string.
